build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/MrjarPlugin.java

```diff
@@ -56,24 +59,38 @@ public void apply(Project project) {
         var javaExtension = project.getExtensions().getByType(JavaPluginExtension.class);
 
         var srcDir = project.getProjectDir().toPath().resolve("src");
+        List<Integer> mainVersions = new ArrayList<>();
         try (var subdirStream = Files.list(srcDir)) {
             for (Path sourceset : subdirStream.toList()) {
                 assert Files.isDirectory(sourceset);
                 String sourcesetName = sourceset.getFileName().toString();
                 Matcher sourcesetMatcher = MRJAR_SOURCESET_PATTERN.matcher(sourcesetName);
                 if (sourcesetMatcher.matches()) {
-                    int javaVersion = Integer.parseInt(sourcesetMatcher.group(1));
-                    addMrjarSourceset(project, javaExtension, sourcesetName, javaVersion);
+                    mainVersions.add(Integer.parseInt(sourcesetMatcher.group(1)));
                 }
             }
         } catch (IOException e) {
             throw new UncheckedIOException(e);
         }
+
+        Collections.sort(mainVersions);
+        String parentSourceSetName = SourceSet.MAIN_SOURCE_SET_NAME;
+        for (int javaVersion : mainVersions) {
+            String sourcesetName = "main" + javaVersion;
+            addMrjarSourceset(project, javaExtension, sourcesetName, parentSourceSetName, javaVersion);
+            parentSourceSetName = sourcesetName; // the next source set will extend this one
+        }
     }
 
-    private void addMrjarSourceset(Project project, JavaPluginExtension javaExtension, String sourcesetName, int javaVersion) {
+    private void addMrjarSourceset(
+        Project project,
+        JavaPluginExtension javaExtension,
+        String sourcesetName,
+        String parentSourceSetName,
+        int javaVersion
+    ) {
         SourceSet sourceSet = javaExtension.getSourceSets().maybeCreate(sourcesetName);
-        GradleUtils.extendSourceSet(project, SourceSet.MAIN_SOURCE_SET_NAME, sourcesetName);
+        GradleUtils.extendSourceSet(project, parentSourceSetName, sourcesetName);
```

Comment:
I don't think this is transitive. That is, if you have `main`, `22` and `23` then `23` will only see output from `22` but not `main` which might cause issues. I believe you'll need to explicitly call this for all earlier versions.

Reply:
The docs for extendsFrom say:

> The artifacts of the super configurations are also available in this configuration.

I also setup a dummy test with additional source sets, and it seemed to work (main classes were available in main22).

Reply:
Indeed `configuration.extendsFrom` is transitive but we are also adding the parent source set output to the classpath but we don't recurse that bit.

> I also setup a dummy test with additional source sets, and it seemed to work (main classes were available in main22).

Was there a `main21` in this scenario?

Reply:
There was. However, I recreated it and found you are right. What's confusing is IntelliJ thinks it's ok, but then at runtime it is not.

Reply:
Never trust IntelliJ 😉

Reply:
This should be addressed in [b109491](https://github.com/elastic/elasticsearch/pull/106092/commits/b10949150176e3ab1cfe299f470ac9fe7cdc9d3d)